### PR TITLE
BLOG-009: BlogIndexPage Model and Listing Logic

### LINKED ISSUE #109: BLOG-009: BlogIndexPage Model and Listing Logic
# BLOG.009: BlogIndexPage Model and Listing Logic

**Phase:** 3 - Blog Models + Templates  
**Priority:** P1  
**Estimated Hours:** 9h  
**Dependencies:** BLOG.005

## Pre-Implementation

**Branch from issue:**
```bash
git checkout develop
git pull origin develop
git checkout -b feature/BLOG-009-blog-index-page
```

## Objective

Create the `BlogIndexPage` Wagtail page model that lists blog posts with pagination and category filtering. This is the main blog landing page (e.g., `/blog/`).

## References

- Implementation Plan: `@/home/mark/workspaces/sum-platform/docs/dev/master-docs/IMPLEMENTATION-PLAN-BLOG-DYNAMICFORMS-v1.md:332-358`
- Wagtail Page Models: https://docs.wagtail.org/en/stable/topics/pages.html
- Pagination: https://docs.djangoproject.com/en/stable/topics/pagination/
- Existing Pages: `core/sum_core/pages/`
- Repository Guidelines: `@/home/mark/workspaces/sum-platform/AGENTS.md`

### Location
- **File:** `core/sum_core/pages/blog.py` (add to existing file with Category)

### Model Definition

```python
from wagtail.models import Page
from wagtail.fields import RichTextField
from wagtail.admin.panels import FieldPanel
from django.core.paginator import Paginator, EmptyPage, PageNotAnInteger
from django.db import models


class BlogIndexPage(Page):
    """
    Blog listing page - displays all blog posts with pagination and filtering.
    
    URL: /blog/
    Template: blog_index_page.html
    """
    
    intro = RichTextField(
        blank=True,
        help_text="Optional intro text displayed above post listing"
    )
    posts_per_page = models.IntegerField(
        default=10,
        help_text="Number of posts to display per page"
    )
    
    content_panels = Page.content_panels + [
        FieldPanel('intro'),
        FieldPanel('posts_per_page'),
    ]
    
    # Page type constraints
    max_count = 1  # Only one blog index per site
    parent_page_types = ['pages.HomePage']  # Or root
    subpage_types = ['pages.BlogPostPage']
    
    def get_posts(self):
        """Get all live blog posts ordered by published date (newest first)."""
        return BlogPostPage.objects.live().descendant_of(self).order_by('-published_date')
    
    def get_posts_by_category(self, category):
        """Get blog posts filtered by category."""
        return self.get_posts().filter(category=category)
    
    def get_context(self, request, *args, **kwargs):
        """Add pagination and category filtering to template context."""
        context = super().get_context(request, *args, **kwargs)
        
        # Get all posts or filter by category
        posts = self.get_posts()
        category_slug = request.GET.get('category')
        selected_category = None
        
        if category_slug:
            try:
                selected_category = Category.objects.get(slug=category_slug)
                posts = self.get_posts_by_category(selected_category)
            except Category.DoesNotExist:
                pass  # Ignore invalid category, show all posts
        
        # Pagination
        page_num = request.GET.get('page', 1)
        paginator = Paginator(posts, self.posts_per_page)
        
        try:
            paginated_posts = paginator.page(page_num)
        except PageNotAnInteger:
            paginated_posts = paginator.page(1)
        except EmptyPage:
            paginated_posts = paginator.page(paginator.num_pages)
        
        # Add to context
        context['posts'] = paginated_posts
        context['categories'] = Category.objects.all()
        context['selected_category'] = selected_category
        
        return context
```

### SEO Integration

Inherit from existing SEO mixins:
```python
from sum_core.seo.mixins import SeoFieldsMixin, OpenGraphMixin, BreadcrumbMixin

class BlogIndexPage(
    SeoFieldsMixin,
    OpenGraphMixin,
    BreadcrumbMixin,
    Page
):
    # ... rest of model
```

### Admin Configuration

Configure admin panels with appropriate grouping:
- **Content Tab:** title, intro, posts_per_page
- **SEO Tab:** Inherited from SEO mixins

## Implementation Tasks

- [ ] Update `core/sum_core/pages/blog.py` to add BlogIndexPage
- [ ] Import required modules (Page, RichTextField, Paginator, etc.)
- [ ] Inherit from existing SEO mixins
- [ ] Define model fields (intro, posts_per_page)
- [ ] Configure content_panels
- [ ] Set parent_page_types and subpage_types constraints
- [ ] Set max_count = 1 (single blog index per site)
- [ ] Implement `get_posts()` method
- [ ] Implement `get_posts_by_category(category)` method
- [ ] Implement `get_context()` with pagination and filtering
- [ ] Update `core/sum_core/pages/__init__.py` to export BlogIndexPage
- [ ] Create migration: `python manage.py makemigrations sum_core`
- [ ] Write unit tests in `tests/pages/test_blog_index_page.py`:
  - BlogIndexPage creation
  - get_posts() returns correct posts
  - get_posts_by_category() filters correctly
  - Pagination works
  - Category filtering via query param
  - Invalid category handled gracefully
  - Empty page edge cases
  - Parent/subpage constraints enforced

## Acceptance Criteria

- [ ] BlogIndexPage can be created in Wagtail admin
- [ ] Only one BlogIndexPage allowed per site
- [ ] Only BlogPostPage allowed as child
- [ ] get_posts() returns live posts ordered by date
- [ ] Pagination works correctly
- [ ] Category filtering via `?category=slug` works
- [ ] Invalid category gracefully shows all posts
- [ ] Empty and out-of-range page numbers handled
- [ ] SEO fields available in admin
- [ ] Migration runs cleanly
- [ ] Unit tests pass with ≥80% coverage
- [ ] `make lint` passes

## Testing Commands

```bash
# Run unit tests
pytest tests/pages/test_blog_index_page.py -v

# Check coverage
pytest tests/pages/test_blog_index_page.py --cov=core/sum_core/pages/blog --cov-report=term-missing

# Interactive testing
python core/sum_core/test_project/manage.py runserver
# Create BlogIndexPage at /blog/
# Verify pagination and filtering work (once BlogPostPage exists)

# Run linting
make lint
```

## Post-Implementation

**Commit, push, and create PR:**
```bash
git add .
git commit -m "feat(blog): add BlogIndexPage model with pagination

- Create BlogIndexPage Wagtail page type
- Implement pagination (configurable posts per page)
- Support category filtering via query parameter
- Integrate with existing SEO mixins
- Configure parent/child page constraints
- Add comprehensive unit tests and migration

Refs: BLOG.009"

git push origin feature/BLOG-009-blog-index-page

gh pr create \
  --base develop \
  --title "feat(blog): BlogIndexPage model with pagination" \
  --body "Implements BLOG.009 - Blog listing page with filtering.

## Changes
- BlogIndexPage Wagtail page model
- Pagination with configurable posts per page
- Category filtering via query param
- get_posts() and get_posts_by_category() methods
- SEO mixin integration
- Parent/child page constraints
- Unit tests and migration

## Testing
- ✅ Page creation works
- ✅ Pagination functions correctly
- ✅ Category filtering works
- ✅ Edge cases handled (invalid category, empty pages)
- ✅ Constraints enforced
- ✅ Lint checks pass

## Related
- Depends on: BLOG.005
- Blocks on: BLOG.012 (BlogIndexPage template)
- Works with: BLOG.011 (BlogPostPage)"
```

**Monitor CI and resolve review comments:**
```bash
gh pr status
gh pr checks --watch
# Address feedback, resolve conversations
```

## Notes for AI Agents

- BlogIndexPage should be **simple** - listing logic only
- Ensure `descendant_of(self)` used to get children (Wagtail pattern)
- Pagination should handle all edge cases (invalid page, empty, out of range)
- Category filtering is optional - default shows all posts
- Query param approach (`?category=slug`) is simpler than path-based routing
- SEO mixins should already exist - reuse them
- Template will be created in BLOG.012 - model only for now
- Consider adding category post counts to context (optional enhancement)

### PR DESCRIPTION
**Pull request by markashton480**
## Summary
- Add BlogIndexPage intro + posts_per_page fields with pagination and category filtering.
- Improve listing robustness (paginator.get_page, request-less context) and annotate categories with post counts for listing use.
- Enforce per-site singleton validation (remove max_count) with save-time validation that can be bypassed for bulk ops.
- Add additive migration for BlogIndexPage fields.

## Testing
- `PATH=./.venv/bin:$PATH make test`
- `PATH=./.venv/bin:$PATH make lint`

## Deferred / Follow-ups
- https://github.com/markashton480/sum-platform/issues/166 (blog templates)
- https://github.com/markashton480/sum-platform/issues/169 (hero block mixin refactor)
- https://github.com/markashton480/sum-platform/issues/170 (category query strategy)
- https://github.com/markashton480/sum-platform/issues/172 (blog optional enhancements)
- https://github.com/markashton480/sum-platform/issues/176 (DB-level singleton enforcement)

Closes #109

### CODE REVIEW

**Comment by markashton480** [comment]
Thanks for the review! Addressed the items:
- BlogPostPage now uses StreamField with PageStreamBlock, plus featured_image + short_description fields and a has_hero_block helper.
- published_date default updated to date.today.
- Added BlogPostPage field/placement tests and a follow-up migration.

Note: wagtail_default_site fixture comes from tests/conftest.py, so no explicit import needed.

Re-ran: make test, make lint.

File: core/sum_core/pages/migrations/0011_blog_pages.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+                        primary_key=True,
+                        serialize=False,
+                        to="wagtailcore.page",
+                    ),
+                ),
+                (
+                    "meta_title",
+                    models.CharField(
+                        blank=True,
+                        help_text="Optional. If blank, defaults to “{page title} | {site name}”.",
```
[the remark this answers is not in the document]
Comment (markashton480) [reply core/sum_core/pages/migrations/0011_blog_pages.py]: Leaving curly quotes because they originate from SeoFieldsMixin help_text and match existing migrations.

File: core/sum_core/pages/__init__.py [hunk 2]
```diff
@@ -37,4 +45,6 @@ def __getattr__(name: str):
     "ServiceIndexPage",
     "ServicePage",
     "Category",
+    "BlogIndexPage",
```
[the remark this answers is not in the document]
Comment (markashton480) [reply core/sum_core/pages/__init__.py]: BlogIndexPage is provided via lazy __getattr__ import in this module, so the __all__ export is valid.

File: core/sum_core/pages/__init__.py [hunk 3]
```diff
@@ -37,4 +45,6 @@ def __getattr__(name: str):
...
+    "BlogPostPage",
```
[the remark this answers is not in the document]
Comment (markashton480) [reply core/sum_core/pages/__init__.py]: BlogPostPage is provided via lazy __getattr__ import in this module, so the __all__ export is valid.

File: core/sum_core/pages/migrations/0011_blog_pages.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (markashton480) [reply core/sum_core/pages/migrations/0011_blog_pages.py]: Leaving the curly quotes as-is because the strings come from existing mixins/migrations; changing would be cross-cutting without functional benefit.

File: core/sum_core/pages/blog.py [hunk 4]
```diff
@@ -48,6 +57,180 @@ def __str__(self) -> str:
...
+
+    class Meta:
+        verbose_name = "Blog Index Page"
+        verbose_name_plural = "Blog Index Pages"
+
+    def get_posts(self):
+        """Return live BlogPostPage children ordered by published date (newest first)."""
+        return (
+            BlogPostPage.objects.live().descendant_of(self).order_by("-published_date")
+        )
```
[the remark this answers is not in the document]
Comment (markashton480) [reply core/sum_core/pages/blog.py]: Added draft exclusion test for get_posts() in tests/pages/test_blog_index_page.py.

**Comment by markashton480** [comment]
Second pass on the latest Claude feedback:
- Added `db_index=True` on BlogPostPage.category with migration `0012_blogpostpage_category_index.py`.
- Updated `get_posts()` to `select_related("category", "featured_image")`.
- Expanded `get_context()` docstring with query params/context keys.
- Clarified max_count/clean intent via inline comment (global guard + per-site validation).

Deferred follow-ups:
- Templates: https://github.com/markashton480/sum-platform/issues/166
- Hero block mixin refactor: https://github.com/markashton480/sum-platform/issues/169
- Category query strategy/caching: https://github.com/markashton480/sum-platform/issues/170
- Optional enhancements (author, RSS, get_absolute_url/docs): https://github.com/markashton480/sum-platform/issues/172

Re-ran: ./.venv/bin/python -m pytest tests/pages/test_blog_index_page.py tests/pages/test_blog_post_page.py -v.

File: core/sum_core/pages/__init__.py [hunk 5]
```diff
@@ -37,4 +45,6 @@ def __getattr__(name: str):
...
 ]
```
[the remark this answers is not in the document]
Comment (markashton480) [reply core/sum_core/pages/__init__.py]: BlogPostPage is provided via the lazy __getattr__ import in core/sum_core/pages/__init__.py, so the __all__ export is still valid and doesn't need removal.

File: tests/pages/test_blog_post_page.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+                    "floating_card_value": "",
+                },
+            }
+        ]
+    )
+
+    post = BlogPostPage(title="Hero Post", slug="hero-post", body=stream_data)
+    blog_index.add_child(instance=post)
+
+    assert post.has_hero_block is True
```
[the remark this answers is not in the document]
Comment (markashton480) [reply tests/pages/test_blog_post_page.py]: BlogPostPage no longer includes has_hero_block on the rebased branch, so the hero_* test coverage is no longer applicable here. The hero mixin refactor remains deferred in #169 if we reintroduce this logic.

File: core/sum_core/pages/blog.py [hunk 7]
```diff
@@ -77,39 +80,109 @@ class CategorySnippetViewSet(SnippetViewSet):
...
+
+        queryset = BlogIndexPage.objects.all()
+        if site is not None and getattr(site, "root_page", None) is not None:
+            queryset = queryset.descendant_of(site.root_page, inclusive=True)
+
+        if queryset.exclude(pk=self.pk).exists():
+            raise ValidationError(
+                {"title": "Only one BlogIndexPage is allowed per site."}
+            )
+
```
[the remark this answers is not in the document]
Comment (markashton480) [reply core/sum_core/pages/blog.py]: Kept clean() alongside max_count because max_count is global; the clean() check scopes to the site tree when possible so multi-site setups can still have one index per site. It falls back to a global check when the site/root isn't available yet.

File: core/sum_core/pages/blog.py [hunk 8]
```diff
@@ -77,39 +80,109 @@ class CategorySnippetViewSet(SnippetViewSet):
...
+    def get_context(self, request, *args, **kwargs):
+        """Add pagination and category filtering to template context."""
+        context = super().get_context(request, *args, **kwargs)
+
+        posts = self.get_posts()
+        query_params = request.GET if request is not None else {}
```
[the remark this answers is not in the document]
Comment (markashton480) [reply core/sum_core/pages/blog.py]: Keeping the request None guard because unit tests (and some callers) invoke get_context(None). Wagtail passes a request at render time, but this keeps the method safe for test usage.

**Comment by markashton480** [comment]
Second pass on Claude critical/high/medium items:\n- get_context now uses paginator.get_page for empty/invalid page handling.\n- Categories are annotated with post_count (uses the current index's live/public posts) to avoid N+1s; added tests for post counts + empty posts.\n- BlogIndexPage.save() now calls clean() so programmatic saves enforce the singleton check; DB-level per-site constraint isn’t straightforward with Wagtail’s tree, so I opened a deferred issue to investigate (#176).\n- Documented request=None behavior and moved homepage creation into a shared fixture (tests/pages/conftest.py).\n\nRe-ran: PATH=./.venv/bin:/tmp/.tmp0SrfrG:/home/codex_c/.npm-global/lib/node_modules/@openai/codex/vendor/x86_64-unknown-linux-musl/path:/home/codex_c/.npm-global/bin:/home/codex_c/.nvm/versions/node/v22.15.0/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin make test, PATH=./.venv/bin:/tmp/.tmp0SrfrG:/home/codex_c/.npm-global/lib/node_modules/@openai/codex/vendor/x86_64-unknown-linux-musl/path:/home/codex_c/.npm-global/bin:/home/codex_c/.nvm/versions/node/v22.15.0/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin make lint.

File: core/sum_core/pages/blog.py [hunk 9]
```diff
@@ -77,39 +81,124 @@ class CategorySnippetViewSet(SnippetViewSet):
...
+        paginator = Paginator(posts, self.posts_per_page)
+        page_num = query_params.get("page", 1)
+        paginated_posts = paginator.get_page(page_num)
+
+        context["posts"] = paginated_posts
+        context["categories"] = Category.objects.annotate(
+            post_count=Count(
+                "blog_posts",
+                filter=Q(blog_posts__in=all_posts),
+            )
```
[the remark this answers is not in the document]
Comment (markashton480) [reply core/sum_core/pages/blog.py]: Updated the category annotation to avoid an __in subquery: counts now use direct conditions (path/depth/live/public) to match the listing query.

File: core/sum_core/pages/blog.py [hunk 10]
```diff
@@ -77,39 +81,124 @@ class CategorySnippetViewSet(SnippetViewSet):
...
+        if queryset.exclude(pk=self.pk).exists():
+            raise ValidationError(
+                {"title": "Only one BlogIndexPage is allowed per site."}
+            )
+
+    def save(self, *args, **kwargs):
+        """Enforce singleton validation even for programmatic saves."""
+        self.clean()
+        super().save(*args, **kwargs)
+
```
[the remark this answers is not in the document]
Comment (markashton480) [reply core/sum_core/pages/blog.py]: Adjusted save() to allow opting out via clean=False (defaults to validation on save for programmatic creation, but bulk operations can skip it).

File: core/sum_core/pages/blog.py [hunk 11]
```diff
@@ -77,39 +81,124 @@ class CategorySnippetViewSet(SnippetViewSet):
...
+                site = parent.get_site()
+
+        queryset = BlogIndexPage.objects.all()
+        if site is not None and getattr(site, "root_page", None) is not None:
+            queryset = queryset.descendant_of(site.root_page, inclusive=True)
+
+        if queryset.exclude(pk=self.pk).exists():
+            raise ValidationError(
+                {"title": "Only one BlogIndexPage is allowed per site."}
+            )
```
[the remark this answers is not in the document]
Comment (markashton480) [reply core/sum_core/pages/blog.py]: Removed max_count so the site-scoped clean() check is the single source of truth; this allows one index per site in multi-site setups.
